plugins/kubernetes/app/models/kubernetes/build_job_executor.rb

```diff
@@ -0,0 +1,133 @@
+# and write progress to the local job output.
+module Kubernetes
+  class BuildJobExecutor
+    WAIT_FOR_JOB = 5.minutes
+    TICK = 3.seconds
+
+    attr_reader :job
+    delegate :id, to: :job
+
+    def initialize(output, job:)
+      @output = output
+      @job = job
+      @registry = {}
+    end
+
+    def execute!(build, project, tag:, push: false, registry:, tag_as_latest: false)
+      job_log = ""
+      k8s_job = nil
+
+      unless valid_registry_info?(registry)
+        @output.puts "### Registry server should not be empty. Aborting..."
+        return false, job_log
+      end
+
+      @registry = registry
+      contents = File.read(build_job_config_path)
+      k8s_job = Kubernetes::JobConfigFile.new(contents, build_job_config_path).job
+      fill_job_details(k8s_job, build, project, tag: tag, push: push, tag_as_latest: tag_as_latest)
+
+      @output.puts "### Running a remote Docker build job: #{k8s_job[:metadata][:name]}"
+      success, job_log = create_and_wait_for_job(k8s_job, @output)
+
+      if success
+        @output.puts "### Remote build job completed sucessfully"
+      else
+        @output.puts "### Remote build job failed to run or timed out"
+      end
+
+      return success, job_log
+    ensure
+      if k8s_job
+        # Jobs will still be there regardless of their statuses
+        @output.puts "### Cleaning up the remote build job: #{k8s_job[:metadata][:name]}"
+        extension_client.delete_job(k8s_job[:metadata][:name], k8s_job[:metadata][:namespace])
+      end
+    end
+
+    private
+
+    def build_cluster
+      Kubernetes::Cluster.first!
+    end
+
+    def client
+      build_cluster.client
+    end
+
+    def extension_client
+      build_cluster.extension_client
+    end
+
+    def valid_registry_info?(registry)
+      registry[:serveraddress].present?
+    end
+
+    def build_job_config_path
+      ENV['KUBE_BUILD_JOB_FILE'] || File.join(Rails.root, 'plugins', 'kubernetes', 'config', 'build_job.yml')
+    end
+
+    def fill_job_details(k8s_job, build, project, tag:, push: false, tag_as_latest: false)
+      # Fill in some information to easily query the job resource
+      project_name = project.name.downcase.tr('_', '-')
+      labels = { project: project_name, role: "docker-build-job" }
+      k8s_job[:metadata][:name] = "docker-build-#{project_name}-#{SecureRandom.hex(7)}"
+      k8s_job[:metadata][:labels].update(labels)
+      k8s_job[:spec][:template][:metadata][:labels].update(labels)
+
+      # Necessary to query the pod running the job later
+      k8s_job[:spec][:template][:metadata][:labels][:jobName] = k8s_job[:metadata][:name]
+
+      # Pass all necessary information so that remote container can build the image
+      container_params = {
+        env: [{name: 'DOCKER_REGISTRY', value: @registry[:serveraddress] }],
+        args: [
+          project.repository_url, build.git_sha, project.docker_repo,
+          tag, push ? "yes" : "no", tag_as_latest ? "yes" : "no"
```

Comment:
this looks a bit crazy ... maybe parentheses can help  tag, (push ? ...), (tag_as_latest ? ...)